### PR TITLE
BUG: Fix wrong error in df drop with non unique datetime index and invalid keys

### LINKED ISSUE #30399: BUG/ERR: wrong error in DataFrame.drop with non-unique datetime index + invalid keys
Consider this example, where there is a DataFrame with a non-unique DatetimeIndex:

```
In [8]: df = pd.DataFrame(np.random.randn(5, 3), columns=['a', 'b', 'c'], index=pd.date_range("2012", freq='H', periods=5))

In [9]: df = df.iloc[[0, 2, 2, 3]] 

In [10]: df
Out[10]: 
                            a         b         c
2012-01-01 00:00:00 -1.534726 -0.559295  0.207194
2012-01-01 02:00:00 -1.072027  0.376595  0.407512
2012-01-01 02:00:00 -1.072027  0.376595  0.407512
2012-01-01 03:00:00  0.581614  1.782635 -0.678197
```

If you then use `drop` to drop some columns, but forget to specify `columns=` or `axis=1` (so you are actually dropping rows), you get a wrong error and very confusing error message:

```
In [10]: df.drop(['a', 'b'])
...

~/scipy/pandas/pandas/core/indexes/base.py in get_indexer_non_unique(self, target)
   4559             tgt_values = target._ndarray_values
   4560 
-> 4561         indexer, missing = self._engine.get_indexer_non_unique(tgt_values)
   4562         return ensure_platform_int(indexer), missing
   4563 

~/scipy/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_indexer_non_unique()

TypeError: 'NoneType' object is not iterable
```

Tested with pandas 0.25 and pandas master.

### PR DESCRIPTION
**Pull request by fujiaxiang**
- [x] closes #30399
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/multi/test_drop.py [hunk 1]
```diff
@@ -139,3 +139,32 @@ def test_drop_not_lexsorted():
...
+    # define dataframe with unique datetime index
+    df_unique = pd.DataFrame(
+        np.random.randn(5, 3),
+        columns=["a", "b", "c"],
+        index=pd.date_range("2012", freq="H", periods=5),
+    )
+    # create dataframe with non-unique datetime index
+    df_nonunique = df_unique.copy().iloc[[0, 2, 2, 3]]
+
+    try:
```
Comment (WillAyd) [inline pandas/tests/indexes/multi/test_drop.py]: I think the point of this is the raise a `KeyError` for the example specified - can you model the test to use `pytest.raises` instead? Should see other examples of this in the code base

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/multi/test_drop.py (hunk 1 above)
Comment (fujiaxiang) [reply pandas/tests/indexes/multi/test_drop.py]: Have updated the code as requested. Now the test specifically expects KeyError with string "not found in axis" within the message

File: doc/source/whatsnew/v1.0.0.rst [hunk 2]
```diff
@@ -714,6 +714,8 @@ Datetimelike
 - Bug in :func:`pandas.to_datetime` failing for `deques` when using ``cache=True`` (the default) (:issue:`29403`)
 - Bug in :meth:`Series.item` with ``datetime64`` or ``timedelta64`` dtype, :meth:`DatetimeIndex.item`, and :meth:`TimedeltaIndex.item` returning an integer instead of a :class:`Timestamp` or :class:`Timedelta` (:issue:`30175`)
 - Bug in :class:`DatetimeIndex` addition when adding a non-optimized :class:`DateOffset` incorrectly dropping timezone information (:issue:`30336`)
+- Bug in :meth:`pandas.core.indexes.base.Index.get_indexer_non_unique` missing condition on target.is_all_date before trying to convert target to asi8 values, which results in wrong error message when dropping with non-unique datetime index (:issue:`30399`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.0.0.rst]:
```suggestion
- Bug in :meth:`DataFrame.drop` where attempting to drop non-existent values from a DatetimeIndex would yield a confusing error message (:issue:`30399`)
```

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks good minor nit on whatsnew - @jorisvandenbossche care to look?

**Comment by jreback** [comment]
thanks @fujiaxiang

**Pull request merged**